src/projection.jl

```diff
@@ -0,0 +1,77 @@
+using LinearAlgebra: Diagonal, diag
+
+"""
+    projector([T::Type], x)
+
+Returns a `project(dx)` closure which maps `dx` onto type `T`, such that it is the
+same size as `x`. If `T` is not provided, it is assumed to be the type of `x`.
+
+It's necessary to have `x` to ensure that it's possible to project e.g. `AbstractZero`s
+onto `Array`s -- this wouldn't be possible with type information alone because the neither
+`AbstractZero`s nor `T` know what size of `Array` to produce.
+"""
+function projector end
+
+projector(x) = projector(typeof(x), x)
+
+# fallback (structs)
+function projector(::Type{T}, x::T) where T
+    project(dx::T) = dx
+    project(dx::AbstractZero) = zero(x)
+    project(dx::AbstractThunk) = project(unthunk(dx))
+    return project
+end
+
+# Numbers
+function projector(::Type{T}, x::T) where {T<:Real}
+    project(dx::Real) = T(dx)
+    project(dx::Number) = T(real(dx)) # to avoid InexactError
+    project(dx::AbstractZero) = zero(x)
+    project(dx::AbstractThunk) = project(unthunk(dx))
+    return project
+end
+function projector(::Type{T}, x::T) where {T<:Number}
+    project(dx::Number) = T(dx)
+    project(dx::AbstractZero) = zero(x)
+    project(dx::AbstractThunk) = project(unthunk(dx))
+    return project
+end
+
+# Arrays
+function projector(::Type{Array{T, N}}, x::Array{T, N}) where {T, N}
+    sizex = size(x)
+    projT = projector(zero(T))
+    project(dx::Array{T, N}) = dx # identity
+    project(dx::AbstractArray) = project(collect(dx)) # from Diagonal
+    project(dx::Array) = projT.(dx) # from different element type
+    project(dx::AbstractZero) = zeros(T, sizex...)
+    project(dx::AbstractThunk) = project(unthunk(dx))
+    return project
+end
+
+# Tangent
+function projector(::Type{<:Tangent}, x::T) where {T}
+    project(dx) = Tangent{T}(; ((k, getproperty(dx, k)) for k in fieldnames(T))...)
+    return project
+end
+
+# Diagonal
+function projector(::Type{<:Diagonal{<:Any, V}}, x::Diagonal) where {V}
+    projV = projector(V, diag(x))
+    project(dx::AbstractMatrix) = Diagonal(projV(diag(dx)))
+    project(dx::Tangent) = Diagonal(projV(dx.diag))
+    project(dx::AbstractZero) = Diagonal(projV(dx))
+    project(dx::AbstractThunk) = project(unthunk(dx))
+    return project
+end
+
+# Symmetric
+function projector(::Type{<:Symmetric{<:Any, M}}, x::Symmetric) where {M}
+    projM = projector(M, parent(x))
+    uplo = Symbol(x.uplo)
+    project(dx::AbstractMatrix) = Symmetric(projM(dx), uplo)
```

Comment:
I don't think this is right, you need to symmetrise, not merely to apply the wrapper.